File: leetcode160/leetcode160.py
# Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution:
    def getIntersectionNode(self, headA: ListNode, headB: ListNode) -> Optional[ListNode]:
        lenA = 0
        lenB = 0
        curA, curB = headA, headB
        while curA != None:
            curA = curA.next
            lenA += 1
        while curB != None:
            curB = curB.next
            lenB += 1
        curA, curB = headA, headB
        if lenB > lenA:
            lenA, lenB = lenB, lenA
            curA, curB = curB, curA
        gap = lenA - lenB
        while gap != 0:
            curA = curA.next
            gap -= 1
        
        while curA != None:
            if curA == curB:
                return curA
            curA = curA.next
            curB = curB.next
        return None

File: leetcode160/leetcode160.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */

struct ListNode {
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(nullptr) {}
};
#include<bits/stdc++.h>
using namespace std;
class Solution {
public:
    ListNode *getIntersectionNode(ListNode *headA, ListNode *headB) {
        ListNode* curA = headA;
        ListNode* curB = headB;
        int lenA = 0, lenB = 0;
        while(curA != nullptr){
            lenA ++;
            curA = curA->next;
        }
        while (curB != nullpter)
        {
            lenB ++;
            curB = curB -> next;
        }

        curA = headA;
        curB = headB;

        if(lenA < lenB){
            swap(lenA, lenB);
            swap(curA, curB);
        }

        int gap = lenA - lenB;
        while (gap --)
        {
            curA = curA -> next;
        }
        while (curA != nullptr)
        {
            if(curA == curB) return curA;
            curA = curA -> next;
            curB = curB -> next;
        }
        return nullptr;
    }
};